src/main/scala/com/sap/kafka/connect/source/hana/HANASourceConnector.scala

```diff
@@ -73,7 +75,22 @@ class HANASourceConnector extends SourceConnector {
     val noOfTables = tables.size
     var tablecount = 1
 
-    var stmtToFetchPartitions = s"SELECT SCHEMA_NAME, TABLE_NAME, PARTITION FROM SYS.M_CS_PARTITIONS WHERE "
+    var stmtToFetchPartitions = s"SELECT SCHEMA_NAME, TABLE_NAME, "
+
+    // Testing for the existence of `TABLE_PARTITIONS` for newer Hana Version (>= SPS 05), otherwise fall back to `M_CS_PARTITIONS`
+    val stmtToCheckPartitionTable = "SELECT CASE WHEN COUNT(*) > 0 THEN 1 ELSE 0 END AS \"exists\" FROM SYS.OBJECTS WHERE SCHEMA_NAME='SYS' AND OBJECT_NAME = 'TABLE_PARTITIONS';"
+    val checkStmt = connection.createStatement()
+    val checkStmtRs = checkStmt.executeQuery(stmtToCheckPartitionTable)
+
+    checkStmtRs.next()
+    if (checkStmtRs.getInt(1) == 1) {
+      log.info(s"Found `TABLE_PARTITIONS` table to determine the partitions")
+      stmtToFetchPartitions += "PART_ID FROM SYS.TABLE_PARTITIONS WHERE "
+    } else {
+      log.info(s"Falling back to `M_CS_PARTITIONS` table to determine the partitions")
+      stmtToFetchPartitions += "PARTITION FROM SYS.M_CS_PARTITIONS WHERE "
+    }
+
```

Comment:
I don't think this part with stmtToCheckPartitionTable is necessary. As table TABLE_PARTITIONS exists everywhere (even in older HANA SPS systems), stmtToCheckPartitionTable will always return 1 and we will be using TABLE_PARTITIONS anyway. In that case, we could actually replace line 76 directly with the appropriate query for TABLE_PARTITIONS.

So, basically we could use TABLE_PARTITIONS or M_CS_TABLES. The SPS 05 documentation mentions both TABLE_PARTITIONS and M_CS_TABLES as the replacement for M_CS_PARTITIONS. In contrast, the Cloud documentation mentions only TABLE_PARTITIONS as the replacement but M_CS_TABLES is available there. So, I asked the HANA colleagues if there is any particular reason for differentiating these two documentations and if there is any concern in using M_CS_TABLES. I hope to get the answer by tomorrow.

Reply:
Thanks for the insights. Seems like only using `TABLE_PARTITIONS` is the way to go. I'll adjust the code accordingly. Do you think the fact that the cloud instance only has entries in `M_CS_TABLES` (for our tables with PART_ID=0), but not in `TABLE_PARTITIONS` matters? I wonder if that changes if a table is partitoined. Maybe we need to query multiple tables to get the right values?

Reply:
Hi @AndreasKlein,  I got the information that the documentation of TABLE_PARTITIONS's PART_ID is incorrect regarding its describing PART_ID = 0 for non-partitioned tables. TABLE_PARTITIONS does not include non-partitioned tables. As for M_CS_TABLES, this table is used by many applications and it is safe to use it on both HANA 2.0 and Cloud HANA to get some simple partition information like the partition count.

So, I think we should use M_CS_TABLES. You mentioned in your issue ticket that you initially used this table in your code. 

Could you update your PR with using M_CS_TABLES? Thanks.

Reply:
@elakito  done. A quick search showed that `M_CS_TABLES` seems to have been there in early SPS as well. We could drop the exists query and move away from using `M_CS_PARTITIONS` for all versions. Can you confirm that?

Reply:
@AndreasKlein Yes. For M_CS_TABLES, you can expect to have this table everywhere, so you don't need the existence-query for this table.

Reply:
@elakito done.